docs/389ds/design/worker-threads.md

```diff
@@ -0,0 +1,345 @@
+- Providing some monitoring data.
+
+## Current implementation synopsys
+
+--------------------
+
+There is a global mutex and a global condition variable and a global queue
+
+- listener threads push data in in the global queue and wake up the worker threads
+- worker threads loops waiting on the global condition variable for new work
+  then process it.
+- to speed up the queueing and limit the alloc/free a free list element is used
+
+## New implementation
+
+--------------------
+
+### Stacked data
+
+In current implementation there are two stacks to handle:
+
+- the work queue element
+- the operation struct
+  and avoid freeing/allocing them every time.
+
+I instead proposed to:
+
+- keep the operation in the thread context. (Anyway a thread only works on a single operation at a time)
+  - keep a free list of element for the waiting_job list. (That cannot be more than the maximum of connections in the queue so in the worse case the free list will 3*8*64K (i.e around 3 Mb) but I wonder if we should not reduce this list size and let the listening threads sleep for a few ms if there are no more items. ( Anyway if the sleep delay is smaller than the time needed to flush the waiting job queue, no time is wasted.)
+
+### Overload warning (flow control)
+
+A warning is logged if
+
+- The warning has not been logged since some delay (1 minute)
+- The waiting job queue size gets higher than a threshold (100*number of worker threads)
+
+### Producters/consumers algorithm
+
+There is still a global mutex and also three lists:
+
+- waiting_threads
+- busy_threads
+- waiting_jobs
+  Access to these lists are protected by the global mutex lock
+
+When pushing new job the listener threads:
+
+- Get global mutex
+
+- Loop forever:
+  
+  - if waiting_threads is not empty:
+    
+    - Select the first waiting thread
+    - Move the thread out of waiting_threads list into the busy_threads list
+    - Release global mutex
+    - Get thread mutex
+    - Assign job to the thread
+    - Wake up that threads
+    - Release Thread mutex
+    - Abort the loop
+  
+  - else if jobs_free_list is not empty:
+    
+    - Pick an waiting_job list element out of the free list
+    - Set the job in the element
+    - push the element at the end of waiting_job list
+    - Release global mutex
+    - Abort the loop
+  
+  - else:
+    
+    - Release global mutex
+    
+    - if current time > last warning time + warning_delay
+      
+      - Log an user friendly warning about worker threads exhausted and waiting jobs queue full.  (maybe: ***Warning: server may be unresponsive because the threads are exhausted and too many operations have been queued.***)
+      
+      - set last warning time to current time
+    
+    - Sleep 20 ms
+    
+    - Get global mutex
+
+The worker threads while waiting on new work:
+
+- Get the thread mutex
+- Loop forever:
+  - if conn is provided in the thread context
+    - Fill the pblock with conn, op
```

Comment:
My understanding is that 'conn' is the way to 'Assign job to the thread'.

Currently the workers are picking a pblock from the work queue. (with conn and op set). Why no longer assigning a job with a pblock ?

Reply:
You are mistaken:  In current code: 
typedef Connection work_q_item;
so it is connection that are queued and not pblock.

Reply:
Opps you are right. sorry.